File: problem-solving/_lc-700-easy-search-in-a-binary-search-tree/2-iterative.go
/*
https://leetcode.com/problems/search-in-a-binary-search-tree/

Complexity
----------
Time: O(n)
Space: O(1)
*/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func searchBST(root *TreeNode, val int) *TreeNode {
    node := root
    for node != nil {
        if node.Val == val {
            return node
        } else if node.Val < val {
            node = node.Right
        } else {
            node = node.Left
        }
    }
    return nil
}

File: problem-solving/lc-98-medium-validate-binary-search-tree/2-iterative.py
"""
https://leetcode.com/problems/validate-binary-search-tree/

Uses inorder traversal and check if we are getting values in ascending order

Complexity
----------
Time: O(N)
Space: O(N)
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        stack = []
        prev = None
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if prev and prev.val >= root.val:
                return False
            prev = root
            root = root.right
        return True
